Development practices/folder_structure.md

```diff
@@ -0,0 +1,286 @@
+│   ├── build_polyline.rb
+│   ├── corridor_search.rb
+│   ├── create_or_update_parking_places.rb
+│   ├── create_or_update_promiles_locations.rb
+│   └── get_route.rb
+├── provider_occupancies_service.rb
+├── providers
+│   ├── client
+│   │   └── mdm.rb
+│   └── parser
+│       ├── a63_france.rb
+│       └── mdm.rb
+├── request_logs
+│   ├── archive_partition.rb
+│   ├── create_archive.rb
+│   └── create_partition.rb
+├── reservations
+│   ├── booking_id_generator.rb
+│   ├── charger.rb
+│   ├── creator.rb
+│   ├── emails.rb
+│   ├── integrations_caller.rb
+│   ├── mails_sender.rb
+│   ├── notifications_sender.rb
+│   ├── settler.rb
+│   ├── status_setter.rb
+│   ├── updater.rb
+│   ├── users_setter.rb
+│   └── voider.rb
+├── shared
+│   └── filter_suggested.rb
+├── stats_maker.rb
+├── subscriptions
+│   └── actions.rb
+├── tln
+│   └── number_checker.rb
+├── translators
+│   └── google.rb
+├── units.rb
+├── update_user_to_developer.rb
+├── user_loyalty_points.rb
+├── users
+│   ├── actions.rb
+│   ├── creator.rb
+│   ├── destroyer.rb
+│   ├── exporter.rb
+│   ├── o_auth_creator.rb
+│   ├── retrieve_all_data
+│   │   ├── base.rb
+│   │   ├── favorites.rb
+│   │   ├── friends_list.rb
+│   │   ├── notifications.rb
+│   │   ├── personal_information.rb
+│   │   └── reviews.rb
+│   ├── retrieve_all_data.rb
+│   └── settings.rb
+├── vat_checker.rb
+├── version_converter.rb
+└── x_server
+    ├── add_detour_distance.rb
+    ├── add_driving_distance.rb
+    ├── base.rb
+    ├── build_polyline.rb
+    ├── corridor_search.rb
+    ├── get_detour_distance.rb
+    ├── get_driving_distance.rb
+    ├── get_geo_keywords.rb
+    ├── get_location_attributes.rb
+    ├── get_route.rb
+    └── leave_reachable.rb
+```
+
+</details>
+
+At some point this folder became a dumpall folder without rhyme or reason.
+There is also an issue of boundries: there are none. Each class is used from anywhere in the application and that can
+cause headaches.
+
+# Desired structure
+
+Here is a folder structure we should strive for
+
+```
+app
+├── controllers  # Rails 
+├── frontend     # webpacker
+├── domains      # custom
+├── models       # Rails
+├── views        # Rails
+└── web          # custom
+```
+
+Lets first talk about Rails specific folders (`controllers`, `views` and `models`).
+
+`views` folder needs to exist because it is hardcoded in the Rails itself (https://github.com/rails/rails/blob/main/railties/lib/rails/engine.rb#L610)
+
+`controllers` is there because of the way rails finds controller names (it appends `"Controller"` to the resource name) (https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/http/request.rb#L88)
+
+`models` is a folder where we put all our ActiveRecord models. It is technically not necessary but we will talk about it in [Further Improvements](#further-improvements) chapter.
+
+
+# `domains` vs `web` folder
+`web` is for classes used in controller actions: forms, policies, serializers, queries and similar.
```

Comment:
This `web` folder approach, I found it very unintuitive in rails apps.
I would still have queries, serializers, and policies while forms should be part of business logic somewhere in domains folder

Reply:
I see what you are saying, but there is no reason for "Forms" to be just a part of domains OR web folder. you can really have both.
Have a form for a domain specific business logic that you might want to reuse everywhere, and the have a route specific form that reuses the domain one.

Reply:
Not sure if I wrote it clear enough so I will rephrase it:

- I would not have web folder at all
- queries, serializers, and policies I would have in `app/serializers`,  `app/queries` and `app/policies` folders
- forms should be part of the business domain, for example, we create users but we can have different user types, so user types should be part of business logic and should not be part or in the scope of place where we use it.